### ISSUE TITLE
(PRIMA) Passing array in abs and comparing not supported

### ISSUE THREAD

**Issue by jinangshah21** [lfortran/lfortran#5285]
Towards #4397 

MRE : 
```fortran
program bobyqa_exmp
real :: A(2) = [2,2]
logical :: B(2)
B = abs(A) > maxval(abs(A))
end program bobyqa_exmp
```

LFortran Error : 
```console
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/mnt/c/lfortran/src/bin/lfortran", in _start
  File "/build/glibc-LcI20x/glibc-2.31/csu/../csu/libc-start.c", line 308, in __libc_start_main
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2677, in main
    return main_app(argc, argv);
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2619, in main_app(int, char**)
    err = compile_src_to_object_file(arg_file, tmp_o, false,
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 1045, in (anonymous namespace)::compile_src_to_object_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::CompilerOptions&, LCompilers::PassManager&)
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/mnt/c/lfortran/src/lfortran/fortran_evaluator.cpp", line 377, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 10317, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass_manager.apply_passes(al, &asr, co.po, diagnostics);
  File "/mnt/c/lfortran/src/libasr/pass/pass_manager.h", line 401, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/mnt/c/lfortran/src/libasr/pass/pass_manager.h", line 178, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    _passes_db[passes[i]](al, *asr, pass_options);
  File "/mnt/c/lfortran/src/libasr/pass/intrinsic_function.cpp", line 521, in LCompilers::pass_replace_intrinsic_function(Allocator&, LCompilers::ASR::TranslationUnit_t&, LCompilers::PassOptions const&)
    v.visit_TranslationUnit(unit);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 49583, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    this->visit_symbol(*a.second);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5189, in LCompilers::ASR::BaseVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 4900, in void LCompilers::ASR::visit_symbol_t<LCompilers::ReplaceIntrinsicFunctionsVisitor>(LCompilers::ASR::symbol_t const&, LCompilers::ReplaceIntrinsicFunctionsVisitor&)
    case symbolType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 49594, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_Program(LCompilers::ASR::Program_t const&)
    self().transform_stmts(xx.m_body, xx.n_body);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 49576, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::transform_stmts(LCompilers::ASR::stmt_t**&, unsigned long&)
    self().visit_stmt(*m_body[i]);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5206, in LCompilers::ASR::BaseVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 4926, in void LCompilers::ASR::visit_stmt_t<LCompilers::ReplaceIntrinsicFunctionsVisitor>(LCompilers::ASR::stmt_t const&, LCompilers::ReplaceIntrinsicFunctionsVisitor&)
    case stmtType::Assignment: { v.visit_Assignment((const Assignment_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 49819, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_Assignment(LCompilers::ASR::Assignment_t const&)
    self().call_replacer();
  File "/mnt/c/lfortran/src/libasr/pass/intrinsic_function.cpp", line 173, in LCompilers::ReplaceIntrinsicFunctionsVisitor::call_replacer()
    replacer.replace_expr(*current_expr);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 48534, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_IntrinsicElementalFunction(down_cast<ASR::IntrinsicElementalFunction_t>(x));
  File "/mnt/c/lfortran/src/libasr/pass/intrinsic_function.cpp", line 60, in LCompilers::ReplaceIntrinsicFunctions::replace_IntrinsicElementalFunction(LCompilers::ASR::IntrinsicElementalFunction_t*)
    replace_expr(x->m_args[i]);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 48538, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_IntrinsicArrayFunction(down_cast<ASR::IntrinsicArrayFunction_t>(x));
  File "/mnt/c/lfortran/src/libasr/pass/intrinsic_function.cpp", line 113, in LCompilers::ReplaceIntrinsicFunctions::replace_IntrinsicArrayFunction(LCompilers::ASR::IntrinsicArrayFunction_t*)
    replace_expr(x->m_args[i]);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 48862, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_ArrayPhysicalCast(down_cast<ASR::ArrayPhysicalCast_t>(x));
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 48074, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_ArrayPhysicalCast(LCompilers::ASR::ArrayPhysicalCast_t*)
    self().replace_expr(x->m_arg);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 48534, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_IntrinsicElementalFunction(down_cast<ASR::IntrinsicElementalFunction_t>(x));
  File "/mnt/c/lfortran/src/libasr/pass/intrinsic_function.cpp", line 90, in LCompilers::ReplaceIntrinsicFunctions::replace_IntrinsicElementalFunction(LCompilers::ASR::IntrinsicElementalFunction_t*)
    ASR::expr_t* current_expr_ = instantiate_function(al, x->base.base.loc,
  File "/mnt/c/lfortran/src/libasr/../libasr/pass/intrinsic_functions.h", line 855, in LCompilers::ASRUtils::Abs::instantiate_Abs(Allocator&, LCompilers::Location const&, LCompilers::SymbolTable*, LCompilers::Vec<LCompilers::ASR::ttype_t*>&, LCompilers::ASR::ttype_t*, LCompilers::Vec<LCompilers::ASR::call_arg_t>&, long)
    body.push_back(al, b.If(b.GtE(args[0], b.f_t(0, arg_types[0])), {
LCompilersException: Expression type, f32[2] not yet supported
```
Works with GFortran

**Comment by AnshMehta1** [comment]
I will work on this and try it out.

**Comment by tanay-man** [comment]
I think an even simpler MRE would be:
```
program bobyqa_exmp

 print * , [-2, 3] > 2
   
end program bobyqa_exmp
```

**Comment by jinangshah21** [comment]
> I think an even simpler MRE would be:
> 
> ```
> program bobyqa_exmp
> 
>  print * , [-2, 3] > 2
>    
> end program bobyqa_exmp
> ```

This is working fine. From Stackrace issue appears to be in abs() function.

**Comment by AnshMehta1** [comment]
> I think an even simpler MRE would be:
> 
> ```
> program bobyqa_exmp
> 
>  print * , [-2, 3] > 2
>    
> end program bobyqa_exmp
> ```

Actulally , this MRE is working
```
(lf) ansh_mehta@DESKTOP-VJGRLON:/mnt/d/lfortran2/lfortran$ ./src/bin/lfortran try.f90
F    T
```

**Comment by AnshMehta1** [comment]
Actually , LHS rank is 1 and RHS rank is 0 in B = abs(A) > maxval(abs(A).
So , compare function makes RHS rank 1 by ArrayPhysicalCast and this thing is getting passed into abs which causes issue as abs works only for rank 0.

**Comment by Pranavchiku** [comment]
Do you think ASR generated for that node is incorrect? Post the ASR corresponding to `abs(A) > maxval(abs(A))` and then we can debug it.

**Comment by AnshMehta1** [comment]
ASR 

```
(lf) ansh_mehta@DESKTOP-VJGRLON:/mnt/d/lfortran2/lfortran$ ./src/bin/lfortran --show-asr try.f90
(TranslationUnit
    (SymbolTable
        1
        {
            bobyqa_exmp:
                (Program
                    (SymbolTable
                        2
                        {
                            a:
                                (Variable
                                    2
                                    a
                                    []
                                    Local
                                    (ArrayConstant
                                        8
                                        [2, 2]
                                        (Array
                                            (Integer 4)
                                            [((IntegerConstant 1 (Integer 4) Decimal)
                                            (IntegerConstant 2 (Integer 4) Decimal))]
                                            FixedSizeArray
                                        )
                                        ColMajor
                                    )
                                    (ArrayConstant
                                        8
                                        [2, 2]
                                        (Array
                                            (Integer 4)
                                            [((IntegerConstant 1 (Integer 4) Decimal)
                                            (IntegerConstant 2 (Integer 4) Decimal))]
                                            FixedSizeArray
                                        )
                                        ColMajor
                                    )
                                    Save
                                    (Array
                                        (Integer 4)
                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                        (IntegerConstant 2 (Integer 4) Decimal))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            b:
                                (Variable
                                    2
                                    b
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Array
                                        (Logical 4)
                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                        (IntegerConstant 2 (Integer 4) Decimal))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    bobyqa_exmp
                    []
                    [(Assignment
                        (Var 2 b)
                        (IntegerCompare
                            (Var 2 a)
                            Gt
                            (IntrinsicArrayFunction
                                MaxVal
                                [(ArrayPhysicalCast
                                    (IntrinsicElementalFunction
                                        Abs
                                        [(ArrayPhysicalCast
                                            (Var 2 a)
                                            FixedSizeArray
                                            DescriptorArray
                                            (Array
                                                (Integer 4)
                                                [((IntegerConstant 1 (Integer 4) Decimal)
                                                (IntegerConstant 2 (Integer 4) Decimal))]
                                                DescriptorArray
                                            )
                                            ()
                                        )]
                                        0
                                        (Array
                                            (Integer 4)
                                            [((IntegerConstant 1 (Integer 4) Decimal)
                                            (IntegerConstant 2 (Integer 4) Decimal))]
                                            FixedSizeArray
                                        )
                                        ()
                                    )
                                    FixedSizeArray
                                    DescriptorArray
                                    (Array
                                        (Integer 4)
                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                        (IntegerConstant 2 (Integer 4) Decimal))]
                                        DescriptorArray
                                    )
                                    ()
                                )]
                                0
                                (Integer 4)
                                ()
                            )
                            (Array
                                (Logical 4)
                                [((IntegerConstant 1 (Integer 4) Decimal)
                                (IntegerConstant 2 (Integer 4) Decimal))]
                                FixedSizeArray
                            )
                            ()
                        )
                        ()
                    )]
                )
        })
    []
)
```

**Comment by Pranavchiku** [comment]
Let's try to understand what is happening, take the following working example:

```fortran
program bobyqa_exmp
real :: A(2) = [2,2]
real :: B(2)
B = abs(A)
end program bobyqa_exmp
```

Now the ASR corresponding to `B = abs(A)` is:

```clojure
                    (Assignment
                        (Var 2 b)
                        (IntrinsicElementalFunction
                            Abs
                            [(ArrayPhysicalCast
                                (Var 2 a)
                                FixedSizeArray
                                DescriptorArray
                                (Array
                                    (Real 4)
                                    [((IntegerConstant 1 (Integer 4) Decimal)
                                    (IntegerConstant 2 (Integer 4) Decimal))]
                                    DescriptorArray
                                )
                                ()
                            )]
                            0
                            (Array
                                (Real 4)
                                [((IntegerConstant 1 (Integer 4) Decimal)
                                (IntegerConstant 2 (Integer 4) Decimal))]
                                FixedSizeArray
                            )
                            ()
                        )
                        ()
                    )
```

If we dump all passes i.e. `--dump-all-passes` and observe the output of `array_op` pass doing  `cat pass_13_array_op.clj`, you won't find the above given node, it is there in all passes before it, so `array_op` does something. It converts that to the following `DoLoop`

```clojure
                    (DoLoop
                        ()
                        ((Var 2 __1_t)
                        (ArrayBound
                            (Var 2 b)
                            (IntegerConstant 1 (Integer 4) Decimal)
                            (Integer 4)
                            LBound
                            ()
                        )
                        (ArrayBound
                            (Var 2 b)
                            (IntegerConstant 1 (Integer 4) Decimal)
                            (Integer 4)
                            UBound
                            ()
                        )
                        ())
                        [(Assignment
                            (ArrayItem
                                (Var 2 b)
                                [(()
                                (Var 2 __1_t)
                                ())]
                                (Real 4)
                                RowMajor
                                ()
                            )
                            (IntrinsicElementalFunction
                                Abs
                                [(ArrayItem
                                    (Var 2 a)
                                    [(()
                                    (Var 2 __1_v)
                                    ())]
                                    (Real 4)
                                    RowMajor
                                    ()
                                )]
                                0
                                (Real 4)
                                ()
                            )
                            ()
                        )
                        (Assignment
                            (Var 2 __1_v)
                            (IntegerBinOp
                                (Var 2 __1_v)
                                Add
                                (IntegerConstant 1 (Integer 4) Decimal)
                                (Integer 4)
                                ()
                            )
                            ()
                        )]
                        []
                    )
```

Simply it converts the fortran code to ( which you can obtain by `--dump-all-passes-fortran`)

```fortran
program bobyqa_exmp
real :: A(2) = [2,2]
real :: B(2)
integer :: i
! B = abs(A)
do i = 1, 2
    B(i) = abs(A(i))
end do
end program bobyqa_exmp
```

Now, as `abs` is an `IntrinsicElementalFunction` and not an `IntrinsicArrayFunction` it expects scalars and not vectors. Which is why the `DoLoop` is created. 

For the original code, let's `--dump-all-passes`, the `DoLoop` gets created as shown:

```clojure
                    (Assignment
                        (Var 2 __libasr_created_scalar_auxiliary_variable)
                        (IntrinsicArrayFunction
                            MaxVal
                            [(ArrayPhysicalCast
                                (IntrinsicElementalFunction
                                    Abs
                                    [(ArrayPhysicalCast
                                        (Var 2 a)
                                        FixedSizeArray
                                        DescriptorArray
                                        (Array
                                            (Real 4)
                                            [((IntegerConstant 1 (Integer 4) Decimal)
                                            (IntegerConstant 2 (Integer 4) Decimal))]
                                            DescriptorArray
                                        )
                                        ()
                                    )]
                                    0
                                    (Array
                                        (Real 4)
                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                        (IntegerConstant 2 (Integer 4) Decimal))]
                                        FixedSizeArray
                                    )
                                    ()
                                )
                                FixedSizeArray
                                DescriptorArray
                                (Array
                                    (Real 4)
                                    [((IntegerConstant 1 (Integer 4) Decimal)
                                    (IntegerConstant 2 (Integer 4) Decimal))]
                                    DescriptorArray
                                )
                                ()
                            )]
                            0
                            (Real 4)
                            ()
                        )
                        ()
                    )
                    (Assignment
                        (Var 2 __1_v)
                        (ArrayBound
                            (Var 2 a)
                            (IntegerConstant 1 (Integer 4) Decimal)
                            (Integer 4)
                            LBound
                            ()
                        )
                        ()
                    )
                    (DoLoop
                        ()
                        ((Var 2 __1_t)
                        (ArrayBound
                            (Var 2 b)
                            (IntegerConstant 1 (Integer 4) Decimal)
                            (Integer 4)
                            LBound
                            ()
                        )
                        (ArrayBound
                            (Var 2 b)
                            (IntegerConstant 1 (Integer 4) Decimal)
                            (Integer 4)
                            UBound
                            ()
                        )
                        ())
                        [(Assignment
                            (ArrayItem
                                (Var 2 b)
                                [(()
                                (Var 2 __1_t)
                                ())]
                                (Logical 4)
                                RowMajor
                                ()
                            )
                            (RealCompare
                                (ArrayItem
                                    (Var 2 a)
                                    [(()
                                    (Var 2 __1_v)
                                    ())]
                                    (Real 4)
                                    RowMajor
                                    ()
                                )
                                Gt
                                (Var 2 __libasr_created_scalar_auxiliary_variable)
                                (Logical 4)
                                ()
                            )
                            ()
                        )
                        (Assignment
                            (Var 2 __1_v)
                            (IntegerBinOp
                                (Var 2 __1_v)
                                Add
                                (IntegerConstant 1 (Integer 4) Decimal)
                                (Integer 4)
                                ()
                            )
                            ()
                        )]
                        []
                    )
```

But the node to observe is what is inside ` (Var 2 __libasr_created_scalar_auxiliary_variable)`. Which comes by looking the `DoLoop` statements.

```clojure
                    (Assignment
                        (Var 2 __libasr_created_scalar_auxiliary_variable)
                        (IntrinsicArrayFunction
                            MaxVal
                            [(ArrayPhysicalCast
                                (IntrinsicElementalFunction
                                    Abs-----------------------------Intrinsic Abs
                                    [(ArrayPhysicalCast
                                        (Var 2 a)
                                        FixedSizeArray
                                        DescriptorArray
                                        (Array ------------------------- Passed Array which should not be the case
                                            (Real 4)
                                            [((IntegerConstant 1 (Integer 4) Decimal)
                                            (IntegerConstant 2 (Integer 4) Decimal))]
                                            DescriptorArray
                                        )
                                        ()
                                    )]
                                    0
                                    (Array
                                        (Real 4)
                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                        (IntegerConstant 2 (Integer 4) Decimal))]
                                        FixedSizeArray
                                    )
                                    ()
                                )
                                FixedSizeArray
                                DescriptorArray
                                (Array
                                    (Real 4)
                                    [((IntegerConstant 1 (Integer 4) Decimal)
                                    (IntegerConstant 2 (Integer 4) Decimal))]
                                    DescriptorArray
                                )
                                ()
                            )]
                            0
                            (Real 4)
                            ()
                        )
                        ()
                    )
```

This means when we go to `replace_IntrinsicArrayFunction` of ( maxval ) we do not visit it's argument. Which can be fixed by the following diff:

```diff
diff --git a/src/libasr/pass/array_op.cpp b/src/libasr/pass/array_op.cpp
index f436bdec0..4cd9ac3e6 100644
--- a/src/libasr/pass/array_op.cpp
+++ b/src/libasr/pass/array_op.cpp
@@ -1476,12 +1476,12 @@ class ReplaceArrayOp: public ASR::BaseExprReplacer<ReplaceArrayOp> {
 
     void replace_IntrinsicArrayFunction(ASR::IntrinsicArrayFunction_t* x) {
         if(!ASRUtils::IntrinsicArrayFunctionRegistry::is_elemental(x->m_arr_intrinsic_id)) {
-            // ASR::BaseExprReplacer<ReplaceArrayOp>::replace_IntrinsicArrayFunction(x);
             if( op_expr == &(x->base) ) {
                 op_dims = nullptr;
                 op_n_dims = ASRUtils::extract_dimensions_from_ttype(
                     ASRUtils::expr_type(*current_expr), op_dims);
             }
+            ASR::BaseExprReplacer<ReplaceArrayOp>::replace_IntrinsicArrayFunction(x);
             return ;
         }
         replace_intrinsic_function(x);

```

**Comment by AnshMehta1** [comment]
Will work on this on the basis of ASR .

**Comment by Pranavchiku** [comment]
Here is the diff to fix the issue:
```diff
diff --git a/src/libasr/pass/array_op.cpp b/src/libasr/pass/array_op.cpp
index f436bdec0..444b42faf 100644
--- a/src/libasr/pass/array_op.cpp
+++ b/src/libasr/pass/array_op.cpp
@@ -1475,16 +1475,30 @@ class ReplaceArrayOp: public ASR::BaseExprReplacer<ReplaceArrayOp> {
     }
 
     void replace_IntrinsicArrayFunction(ASR::IntrinsicArrayFunction_t* x) {
-        if(!ASRUtils::IntrinsicArrayFunctionRegistry::is_elemental(x->m_arr_intrinsic_id)) {
-            // ASR::BaseExprReplacer<ReplaceArrayOp>::replace_IntrinsicArrayFunction(x);
-            if( op_expr == &(x->base) ) {
-                op_dims = nullptr;
-                op_n_dims = ASRUtils::extract_dimensions_from_ttype(
-                    ASRUtils::expr_type(*current_expr), op_dims);
+        if( op_expr == &(x->base) ) {
+            op_dims = nullptr;
+            op_n_dims = ASRUtils::extract_dimensions_from_ttype(
+                ASRUtils::expr_type(*current_expr), op_dims);
+        }
+        for( size_t i = 0; i < x->n_args; i++ ){
+            ASR::expr_t* arg = x->m_args[i];
+            if ( ASR::is_a<ASR::ArrayPhysicalCast_t>(*arg) ) {
+                arg = ASR::down_cast<ASR::ArrayPhysicalCast_t>(arg)->m_arg;
+            }
+
+            if ( ASR::is_a<ASR::IntrinsicElementalFunction_t>(*arg)) {
+                ASR::expr_t** current_expr_copy_9 = current_expr;
+                current_expr = &(x->m_args[i]);
+                ASR::dimension_t* op_dims_copy = op_dims;
+                size_t op_n_dims_copy = op_n_dims;
+                self().replace_expr(x->m_args[i]);
+                x->m_args[i] = *current_expr;
+                op_dims = op_dims_copy;
+                op_n_dims = op_n_dims_copy;
+                current_expr = current_expr_copy_9;
             }
-            return ;
         }
-        replace_intrinsic_function(x);
+        return ;
     }
 
```

**Comment by AnshMehta1** [comment]
Surely, will work on this changes at night.

**Comment by Pranavchiku** [comment]
Fixed by #5336
